pkg/util/syncutil/singleflight/singleflight.go

```diff
@@ -251,7 +251,13 @@ func (c *call) result(ctx context.Context, leader bool) Result {
 		case <-c.c:
 		case <-ctx.Done():
 			op := fmt.Sprintf("%s:%s", c.opName, c.key)
-			if !leader {
+			if leader {
```

Comment:
I this that this should only happen if `InheritCancellation` is true. Another thing that might be good to do here is to log if we wait a long time here. Maybe:

```
var timer timeutil.Timer
defer timer.Stop()
const slowLogTimeout = time.Second
timer.Reset(slowLogTimeout)
start := timeutil.Now()
select {
case <-c.c:
    return
case <-timer.C:
    timer.Read = true
    log.Infof(ctx, "have been waiting for singleflight %s:%s for %v after cancelation", c.opName, c.key, timeutil.Since(start))
}
<-c.c
```

Reply:
I don't know. I think I'd argue instead that the singleflights users which share state and want to exit early are abusing the API. Maybe we should just comment that and not build this into the singleflight package.